assets/app/scripts/services/apiEndpoint_v2.js

```diff
@@ -0,0 +1,311 @@
+    };
+
+    var normalizeResource = function(resource) {
+       if (!resource) {
+        return;
+       }
+
+       // only lowercase the first segment, leaving subresources as-is (some are case-sensitive)
+       var segments = resource.split("/");
+       segments[0] = segments[0].toLowerCase();
+       var normalized = segments.join("/");
+       if (resource !== normalized) {
+         Logger.warn('Non-lower case resource "' + resource + '"');
+       }
+
+       return normalized;
+    };
+
+    // port of restmapper.go#kindToResource
+    var kindToResource = function(kind) {
+      if (!kind) {
+        return "";
+      }
+      var resource = String(kind).toLowerCase();
+      if (resource.endsWith('status')) {
+        resource = resource + 'es';
+      } else if (resource.endsWith('s')) {
+        // no-op
+      } else if (resource.endsWith('y')) {
+        resource = resource.substring(0, resource.length-1) + 'ies';
+      } else {
+        resource = resource + 's';
+      }
+      // make sure it is a known resource
+      if (!apiExistsFor(resource)) {
+        Logger.warn('Unknown resource "' + resource + '"');
+        return undefined;
+      }
+      return resource;
+    };
+
+    // TODO: this is not efficient. make another map & do a lookup,
+    // also use apiVersion (if we need it)
+    var apiExistsFor = function(resource, apiVersion) {
+      var found = false;
+      _.each(apis, function(api) {
+        _.each(api, function(version) {
+          _.each(version, function(res, key) {
+            if(resource === key) {
+              found = true;
+            }
+          });
+        });
+      });
+      return found;
+    };
+
+    // deprecated?
+    var ensureVersion = function(api, preferredAPIVersion) {
+      return preferredAPIVersion || (api && api.defaultVersion);
+    };
+
+    // Bother, URITemplate has no ability to do conditional {+group}, so
+    // templates have to be more complex
+    var API_TEMPLATE = "{protocol}://{+hostPort}{+prefix}/{version}/";
+    var API_GROUP_TEMPLATE = "{protocol}://{+hostPort}{+prefix}/{+group}/{version}/";
+
+    var URL_GET_LIST              = "{resource}{?q*}";
+    var URL_OBJECT                = "{resource}/{name}{/subresource*}{?q*}";
+    var URL_NAMESPACED_GET_LIST   = "namespaces/{namespace}/{resource}{?q*}";
+    var URL_NAMESPACED_OBJECT     = "namespaces/{namespace}/{resource}/{name}{/subresource*}{?q*}";
+
+    var findTemplateFor = function(name, namespace, group) {
+      var base = group ?
+                  API_GROUP_TEMPLATE :
+                  API_TEMPLATE;
+      var rest = name ?
+                  namespace ?
+                    URL_NAMESPACED_OBJECT :
+                    URL_OBJECT :
+                  namespace ?
+                    URL_NAMESPACED_GET_LIST :
+                    URL_GET_LIST;
+      return (base + rest);
+    };
+
+    var protocol = function(isWebsocket) {
+      return isWebsocket ?
```

Comment:
Nice, appreciate all the comments & will get working on these. FWIW on the ternary style, I used to think ternary were always inferior to an if statement, until landing on the 3 line version:
```
question ?
  truthyAnswer :
  falsyAnswer;

// sometimes answer a question w/a question :)
question ?
  truthyQuestion ?
    truthyAnswer :
    falsyAnswer :
  falsyQuestion ?
    truthyAnswer :
    falsyAnswer :
```

I think when you break it into 3 lines its extremely easy to read & nice and terse (w/o needing unnecessary vars).  I wouldn't go any deeper than this though, agree there is a point when its not readable.